spec/models/results/web_response_parser_spec.rb

```diff
@@ -0,0 +1,178 @@
+        {
+          "0" => web_answer_hash(form.c[0].id, value: "A"),
+          "1" => web_answer_hash(form.c[1].id, value: "B"),
+          "2" => web_answer_hash(form.c[2].id, value: "C")
+        }
+      end
+
+      it "builds tree with three answers" do
+        input = ActionController::Parameters.new(data)
+        tree = Results::WebResponseParser.new.parse(input)
+        expect_root(tree, form)
+        expect_children(tree, %w[Answer Answer Answer], form.c.map(&:id), %w[A B C])
+      end
+    end
+
+    context "with one irrelevant answer" do
+      let(:answers) do
+        {
+          "0" => web_answer_hash(form.c[0].id, value: "A"),
+          "1" => web_answer_hash(form.c[1].id, {value: "B"}, relevant: "false"),
+          "2" => web_answer_hash(form.c[2].id, value: "C")
+        }
+      end
+
+      it "builds tree with two answers" do
+        input = ActionController::Parameters.new(data)
+        tree = Results::WebResponseParser.new.parse(input)
+        expect_root(tree, form)
+        expect_children(tree, %w[Answer Answer], [form.c[0].id, form.c[2].id], %w[A C])
+      end
+    end
+
+    context "with one destroyed answer" do
+      let(:answers) do
+        {
+          "0" => web_answer_hash(form.c[0].id, value: "A"),
+          "1" => web_answer_hash(form.c[1].id, {value: "B"}, destroy: "true"),
+          "2" => web_answer_hash(form.c[2].id, value: "C")
+        }
+      end
+
+      it "builds tree with two answers" do
+        input = ActionController::Parameters.new(data)
+        tree = Results::WebResponseParser.new.parse(input)
+        expect_root(tree, form)
+        expect_children(tree, %w[Answer Answer], [form.c[0].id, form.c[2].id], %w[A C])
+      end
+    end
+  end
+
+  context "response with an answer set" do
+    let(:question_types) { %w[text multilevel_select_one text] }
+    let(:data) { {root: web_answer_group_hash(form.root_group.id, answers)} }
+    let(:plant) { form.c[1].option_set.sorted_children[1].id }
+    let(:oak) { form.c[1].option_set.sorted_children[1].sorted_children[1].id }
+    let(:answers) do
+      {
+        "0" => web_answer_hash(form.c[0].id, value: "A"),
+        "1" => {
+          id: "",
+          type: "AnswerSet",
+          questioning_id: form.c[1].id,
+          relevant: "true",
+          children: {
+            "0" => web_answer_hash(form.c[1].id, option_node_id: plant),
+            "1" => web_answer_hash(form.c[1].id, option_node_id: oak)
+          }
+        },
+        "2" => web_answer_hash(form.c[2].id, value: "D")
+      }
+    end
+
+    it "builds tree with answer set" do
+      tree = Results::WebResponseParser.new.parse(input)
+      expect_root(tree, form)
+      expect_children(tree, %w[Answer AnswerSet Answer], form.c.map(&:id), ["A", nil, "D"])
+      expect_children(tree.c[1], %w[Answer Answer], [form.c[1].id, form.c[1].id], %w[Plant Oak])
+    end
+  end
+
+  context "builds tree for response with select multiple answer" do
```

Comment:
I think it would be clearer and more consistent to remove `builds tree for` here.